src/lib/index/content-claims.js

```diff
@@ -0,0 +1,130 @@
+/* global ReadableStream */
+import * as Link from 'multiformats/link'
+import * as raw from 'multiformats/codecs/raw'
+import * as Claims from '@web3-storage/content-claims/client'
+import { MultihashIndexSortedReader } from 'cardex/multihash-index-sorted'
+import { Map as LinkMap } from 'lnmap'
+import { Set as LinkSet } from 'lnset'
+import { CAR_CODE } from '../../constants'
+
+/**
+ * @typedef {import('multiformats').UnknownLink} UnknownLink
+ * @typedef {import('./api').IndexEntry} IndexEntry
+ * @typedef {import('./api').Index} Index
+ */
+
+/** @implements {Index} */
+export class ContentClaimsIndex {
+  /**
+   * Cached index entries.
+   * @type {Map<UnknownLink, IndexEntry>}
+   */
+  #cache
+  /**
+   * CIDs for which we have already read claims.
+   * @type {Set<UnknownLink>}
+   */
+  #seen
+  /**
+   * @type {URL|undefined}
+   */
+  #serviceURL
+
+  /**
+   * @param {{ serviceURL?: URL }} [options]
+   */
+  constructor (options) {
+    this.#cache = new LinkMap()
+    this.#seen = new LinkSet()
```

Comment:
minor: Why separate cache and "seen things" cache? for cids that we have indexes for we store the cid twice? not a big deal, but could store a magic "no index" value in the cache instead.